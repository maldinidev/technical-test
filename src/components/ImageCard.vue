<template>
  <div class="relative group">
    <img :src="image.download_url" :alt="image.author" @load="handleImageLoad" loading="lazy"
      class="w-full h-48 object-cover shadow transition duration-700 ease-in-out transform"
      :class="image.loaded ? 'opacity-100 scale-100' : 'opacity-0 scale-95'" />
    <div
      class="absolute inset-0 flex items-center justify-center bg-black bg-opacity-50 opacity-0 group-hover:opacity-85 duration-700 ease-in-out cursor-pointer"
      @click="removeImage">
      <span class="text-white font-bold">Eliminar</span>
    </div>
  </div>
</template>

<script setup>
// Define the props for the component
const props = defineProps({
  image: {
    type: Object, // Expecting an object with properties like `download_url`, `author`, and `loaded`
    required: true, // This prop is mandatory for the component to work
  },
});

// Define the events the component can emit
const emit = defineEmits(['remove']); // Declaring the 'remove' event that the parent component will listen to

// Function to mark the image as loaded
const handleImageLoad = () => {
  props.image.loaded = true; // Update the `loaded` property of the image prop when the image finishes loading
};

// Function to emit the 'remove' event to the parent
const removeImage = () => {
  emit('remove', props.image); // Emit the 'remove' event with the image object as the payload
};
</script>
