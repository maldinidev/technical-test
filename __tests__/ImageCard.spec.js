import { mount } from '@vue/test-utils';
import ImageCard from '../src/components/ImageCard.vue';

describe('ImageCard.vue', () => {
  const mockImage = {
    id: 1,
    download_url: 'https://picsum.photos/id/1/200/300',
    author: 'Author 1',
    loaded: false,
  };

  it('renderiza la imagen correctamente', () => {
    const wrapper = mount(ImageCard, {
      props: {
        image: mockImage,
      },
    });

    const img = wrapper.find('img');
    expect(img.attributes('src')).toBe(mockImage.download_url);
    expect(img.attributes('alt')).toBe(mockImage.author);
  });

  it('emite el evento "remove" al hacer clic en el área de eliminación', async () => {
    const wrapper = mount(ImageCard, {
      props: {
        image: mockImage,
      },
    });

    const removeOverlay = wrapper.find('.group-hover\\:opacity-85');
    await removeOverlay.trigger('click');

    expect(wrapper.emitted('remove')).toBeTruthy();
    expect(wrapper.emitted('remove')[0]).toEqual([mockImage]);
  });

  it('marca la imagen como cargada al disparar el evento load', async () => {
    const wrapper = mount(ImageCard, {
      props: {
        image: mockImage,
      },
    });

    const img = wrapper.find('img');
    await img.trigger('load');

    expect(mockImage.loaded).toBe(true);
  });

  it('applies the correct classes based on the "loaded" state', async () => {
    const wrapper = mount(ImageCard, {
      props: {
        image: { ...mockImage, loaded: false },
      },
    });

    const img = wrapper.find('img');
    expect(img.classes()).toContain('opacity-0');
    expect(img.classes()).toContain('scale-95');

    // Simulate the image load
    await img.trigger('load');
    expect(img.classes()).toContain('opacity-100');
    expect(img.classes()).toContain('scale-100');
  });

  it('matches the snapshot', () => {
    const wrapper = mount(ImageCard, {
      props: {
        image: mockImage,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
