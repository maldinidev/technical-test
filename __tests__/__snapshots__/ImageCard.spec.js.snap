// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageCard.vue matches the snapshot 1`] = `
"<div class="relative group"><img src="https://picsum.photos/id/1/200/300" alt="Author 1" loading="lazy" class="w-full h-48 object-cover shadow transition duration-700 ease-in-out transform opacity-100 scale-100">
  <div class="absolute inset-0 flex items-center justify-center bg-black bg-opacity-50 opacity-0 group-hover:opacity-85 duration-700 ease-in-out cursor-pointer"><span class="text-white font-bold">Eliminar</span></div>
</div>"
`;
